### PR TITLE
DataGrid.Pdf - BuiltIn calculations/drawing for each 'manual definition' from the gridRect argument only

### PR DESCRIPTION
**Pull request by KuznetsovVN**

### CODE REVIEW

File: js/exporter/jspdf/pdf_grid.js [hunk 1]
```diff
@@ -28,7 +29,7 @@ export class PdfGrid {
         if(isDefined(this._splitByColumns)) {
             this._splitByColumns.forEach((splitByColumn) => {
                 this._currentHorizontalTables.push(
-                    new PdfTable(drawTableBorder, splitByColumn.tableRect)
+                    new PdfTable(drawTableBorder, splitByColumn.tableRect, this._columnWidths.slice(splitByColumn.columnIndex))
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_grid.js]: Here we will pass too many elements from the 'columnWIdths' array: each table will receive width of its columns and column width of all the rest tables.
Whether it's better to sync the length of the passed array with the number of the current table columns?

File: js/exporter/jspdf/pdf_table.js [hunk 2]
```diff
@@ -2,16 +2,35 @@ import { isDefined } from '../../core/utils/type';
...
-    constructor(drawTableBorder, rect) {
+    constructor(drawTableBorder, rect, columnWidths) {
         this.drawTableBorder = drawTableBorder;
         this.rect = rect;
+        this.columnWidths = columnWidths ?? [];
         this.rows = [];
     }
 
-    addRow(cells) {
+    tryGetXPos(cellIndex) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Method does not 'try', it always returns some value.
Let's rename it to getCellLeft/getCellX

File: js/exporter/jspdf/pdf_table.js [hunk 3]
```diff
@@ -2,16 +2,35 @@ import { isDefined } from '../../core/utils/type';
...
+        return this.rect.x + this.columnWidths.slice(0, cellIndex).reduce((a, b) => a + b, 0);
+    }
+
+    tryGetYPos(cellIndex) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Method does not 'try', it always returns some value.
Let's rename it to getCellTop/getCellY

File: js/exporter/jspdf/pdf_table.js [hunk 4]
```diff
@@ -34,6 +53,34 @@ export class PdfTable {
                     currentCell.drawTopBorder = false;
                 }
             }
+
+
+            if(!cells[i].skip) {
+                if(!isDefined(cells[i].rect)) {
+                    cells[i].rect = {};
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Let's mark this member as 'for internal use': cells[i]._rect = {};

File: js/exporter/jspdf/pdf_table.js [hunk 5]
```diff
@@ -2,16 +2,35 @@ import { isDefined } from '../../core/utils/type';
...
-    addRow(cells) {
+    tryGetXPos(cellIndex) {
+        return this.rect.x + this.columnWidths.slice(0, cellIndex).reduce((a, b) => a + b, 0);
+    }
+
+    tryGetYPos(cellIndex) {
+        let currentYPos = this.rect.y;
+        for(let index = 0; index < this.rows.length - 1; index++) {
+            const cell = this.rows[index][cellIndex];
+            if(cell && cell.rect.h) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: There is the 'if(!cells[i].skip)' condition and 'rect' will not be initialized.
How code will manage 'height' in this case?
Whether we can add a test for this case?

File: js/exporter/jspdf/pdf_table.js [hunk 6]
```diff
@@ -34,6 +53,34 @@ export class PdfTable {
                     currentCell.drawTopBorder = false;
                 }
             }
+
+
+            if(!cells[i].skip) {
+                if(!isDefined(cells[i].rect)) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Whether the 'rect' member value can be assigned by other code?

File: js/exporter/jspdf/pdf_table.js [hunk 7]
```diff
@@ -34,6 +53,34 @@ export class PdfTable {
...
+                    cells[i].rect = {};
+                }
+
+                if(!isDefined(cells[i].rect.x)) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js]: Whether the 'x' member value can be assigned by other code?

File: js/exporter/jspdf/pdf_table.js [hunk 8]
```diff
@@ -2,16 +2,35 @@ import { isDefined } from '../../core/utils/type';
...
+        for(let index = 0; index < this.rows.length - 1; index++) {
+            const cell = this.rows[index][cellIndex];
+            if(cell && cell.rect.h) {
+                currentYPos += cell.rect.h;
+            }
+        }
+        return currentYPos;
+    }
+
+    addRow(cells, rowHeight) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_table.js:25]: Whether it is better to check 'cells.legth === this.columnWidths.length' ?

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js [hunk 9]
```diff
@@ -1641,6 +1390,15 @@ QUnit.module('exportDataGrid', moduleConfig, () => {
                 e.drawNewTableFromThisRow.startNewTable = true;
                 e.drawNewTableFromThisRow.addPage = true;
                 e.drawNewTableFromThisRow.tableRect = { x: 10, y: 56, w: 40, h: 54 };
+                e.drawNewTableFromThisRow.splitToTablesByColumns = [{
+                    columnIndex: 1,
+                    drawOnNewPage: true,
+                    tableRect: { x: 11, y: 21, w: 50, h: 54 }
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js]: Let's [x,y] for all tables: 
e.drawNewTableFromThisRow.tableRect ={ x: 10, y: 56, w: 40, h: 54 }; -> '{ x: 12, y: 22, w: 40, h: 54 };'
tableRect: { x: 11, y: 21, w: 50, h: 54 } -> { x: 13, y: 23, w: 50, h: 54 }
tableRect: { x: 12, y: 22, w: 60, h: 54 } -> { x: 14, y: 24, w: 60, h: 54 }

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js [hunk 10]
```diff
@@ -1567,6 +1316,15 @@ QUnit.module('exportDataGrid', moduleConfig, () => {
...
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js]: Let's [x,y] for 3 tables

File: js/exporter/jspdf/pdf_grid.js [hunk 11]
```diff
@@ -15,7 +16,7 @@ export class PdfGrid {
 
     startNewTable(drawTableBorder, firstTableRect, firstTableOnNewPage, splitByColumns) {
         this._currentHorizontalTables = [
-            new PdfTable(drawTableBorder, firstTableRect)
+            new PdfTable(drawTableBorder, firstTableRect, this._columnWidths.slice(0))
```
Comment (IgnatovDan) [inline js/exporter/jspdf/pdf_grid.js]: PdfTable will receive its width for its columns and for all the rest columns.
Whether it's better to sync the length of the passed array with the number of this table columns?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/exporter/jspdf/pdf_grid.js (hunk 1 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_grid.js]: ok

File: js/exporter/jspdf/pdf_table.js (hunk 2 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js]: ok

File: js/exporter/jspdf/pdf_table.js (hunk 3 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js]: ok

File: js/exporter/jspdf/pdf_table.js (hunk 4 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js]: ok

File: js/exporter/jspdf/pdf_table.js (hunk 5 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js]: Cells with cells[i].skip === true are not participate in calculation. Their sizes were merged with master cell ( Cells with attributes: colSpan, rowSpan )

File: js/exporter/jspdf/pdf_table.js (hunk 7 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js]: Currently x,y,w,h may assign in case with band cells

File: js/exporter/jspdf/pdf_table.js (hunk 8 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_table.js:25]: ok

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js (hunk 9 above)
Comment (KuznetsovVN) [reply testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js]: ok

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js (hunk 10 above)
Comment (KuznetsovVN) [reply testing/tests/DevExpress.exporter/jspdfParts/jspdf_v2.dataGrid.tests.js]: ok

File: js/exporter/jspdf/pdf_grid.js (hunk 11 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/pdf_grid.js]: ok

**Review by IgnatovDan — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by novsstation — APPROVED** [review]

**Pull request merged**